plugin/pkg/admission/podplacementrestriction/admission.go

```diff
@@ -0,0 +1,248 @@
+		pluginConfig := readConfig(config)
+		return NewPodTolerationsPlugin(client, pluginConfig.PodTolerationsPluginConfig["clusterDefaultTolerations"]), nil
+	})
+}
+
+var _ = admission.Interface(&podTolerationsPlugin{})
+var _ = admission.WantsInformerFactory(&podTolerationsPlugin{})
+
+type podTolerationsPlugin struct {
+	*admission.Handler
+	client            clientset.Interface
+	namespaceInformer framework.SharedIndexInformer
+	// global default tolerations in a cluster. If a namespace is
+	// not assigned tolerations, it gets this by default.
+	clusterDefaultTolerations []api.Toleration
+}
+
+type pluginConfig struct {
+	PodTolerationsPluginConfig map[string][]api.Toleration
+}
+
+// readConfig reads default value of clusterDefaultTolerations
+// from the file provided with --admission-control-config-file
+// If the file is not supplied, it defaults to "".
+// The format in a file:
+// podTolerationsPluginConfig:
+//  clusterDefaultTolerations:
+//    - Key: key1
+//      Value: value1
+//    - Key: key2
+//      Value: value2
+func readConfig(config io.Reader) *pluginConfig {
+	defaultConfig := &pluginConfig{}
+	if config == nil || reflect.ValueOf(config).IsNil() {
+		return defaultConfig
+	}
+	d := yaml.NewYAMLOrJSONDecoder(config, 4096)
+	for {
+		if err := d.Decode(defaultConfig); err != nil {
+			if err != io.EOF {
+				continue
+			}
+		}
+		break
+	}
+	return defaultConfig
+}
+
+// Admit merges pod's and its namespace tolerations.
+func (p *podTolerationsPlugin) Admit(a admission.Attributes) error {
+	resource := a.GetResource().GroupResource()
+	if resource != api.Resource("pods") {
+		return nil
+	}
+	if a.GetSubresource() != "" {
+		// only run the checks below on pods proper and not subresources
+		return nil
+	}
+
+	obj := a.GetObject()
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		return fmt.Errorf("expected pod but got %s", a.GetKind().Kind)
+	}
+
+	if !p.WaitForReady() {
+		return admission.NewForbidden(a, fmt.Errorf("not yet ready to handle request"))
+	}
+
+	nsName := a.GetNamespace()
+	namespaceObj, exists, err := p.namespaceInformer.GetStore().Get(&api.Namespace{
+		ObjectMeta: api.ObjectMeta{
+			Name:      nsName,
+			Namespace: "",
+		},
+	})
+	if err != nil {
+		return errors.NewInternalError(err)
+	}
+
+	if !exists {
+		// in case of latency in our caches, make a call direct to storage to verify that it truly exists or not
+		namespaceObj, err = p.client.Core().Namespaces().Get(nsName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return err
+			}
+			return errors.NewInternalError(err)
+		}
+	}
+
+	namespace := namespaceObj.(*api.Namespace)
+	nsTolerations, err := p.getNamespaceTolerations(namespace)
+	if err != nil {
+		return err
+	}
+
+	podTolerations, err := api.GetTolerationsFromAnnotations(pod.Annotations)
+	if err != nil {
+		return err
+	}
+
+	if conflictingTolerations(nsTolerations, podTolerations) {
```

Comment:
@sttts yeah I haven't thought much about it, basically replicating what existed with levels. I think will give it a more thought now.